torchtune/utils/_distributed.py

```diff
@@ -318,15 +320,32 @@ def load_from_full_model_state_dict(
             chunk = list(torch.chunk(full_tensor, shard_world_size, dim=0))[shard_rank]
             sharded_param = full_tensor.new_zeros(chunk.size())
             sharded_param[: chunk.size(0)].copy_(chunk)
-            sharded_tensor = DTensor(
-                sharded_param,
-                sharded_meta_param.device_mesh,
-                sharded_meta_param.placements,
-                shape=sharded_meta_param.size(),
-                dtype=sharded_meta_param.dtype,
-                requires_grad=sharded_meta_param.requires_grad,
-                stride=sharded_meta_param.stride(),
-            )
+            # BC-breaking change to DTensor API in https://github.com/pytorch/pytorch/pull/128112
+            if version.parse(torch.__version__) >= version.parse("2.4.0.dev20240606"):
```

Comment:
@wanchaol @weifengpy please let me know if this is the best way to handle this, also open to any other ideas either of you have here.

Reply:
Also regarding a longer-term plan here, are there any plans to make `DTensor` a public API? Or should we consider whether there's an alternative way to do this without directly calling into the `DTensor` API?

Reply:
chatted about this. the long-term solution is to use public DTensor api

```
return DTensor.from_local(
    local_tensor,
    sharding_spec.mesh,
    sharding_spec.placements,
    shape=sharding_spec.shape,
    stride=sharding_spec.stride,
)
```

Reply:
since this is state_dict logic, I think we should just directly use `DTensor.from_local`? `DTensor.__new__` is kinda private API of DTensor

Reply:
One more question here: if using with NF4, does it mean we will have to also add logic for aten.view in ao (based on [this line](https://github.com/pytorch/pytorch/blob/529bedad3e6cbd98387b9e46ce0c0e290ed12ab5/torch/distributed/_tensor/api.py#L151))?

Reply:
good catch. i think so. do you know the args, kwargs after dispatching? if view_as(shape) matches NF4.shape, that’s trivial. otherwise we need a deeper look because the semantics of NF4.view_as(different shape)

Reply:
btw, it does not have to be addressed in this PR, if it needs torchao change

Reply:
Yeah personally I don't know the args or kwargs so need to do a bit more investigation. Will leave this as a fast follow for now

Reply:
I just found pytorch from source does not work.  `torch.__version__` is `2.4.0a0+git....`. My commit hash is before DTensor BC breaking, but It goes into `DTensor(spec)` and complained cc @ebsmothers